### DISCUSSION TITLE
(Partial) Onboarding to an existing cluster and interaction with existing resources

### DISCUSSION THREAD

**Discussion by pinkfloydx33** [fluxcd/flux2#2670] [Q&A] [answered]
I'm interested in bringing Flux into our infrastructure, but I have some questions that aren't clear to me after reading through the docs/issues/etc.

We have two "classifications" of resources in our clusters: those associated *directly* with our applications/services and those that are "part of the infrastructure". The latter group includes things like `nginx`, `certmanager`, `external-dns` as well as various `ClusterRole`, `ClusterIssuer`, `PersistentVolume` and `Secret` resources (amongst others). These are all currently managed via Helm charts, manifests and a manual (though scripted) update process. The former group (the applications) are all internally packaged Helm charts with deployment, gating and approval controlled via Azure Devops.

For the moment, I am only concerned with the second group (the infrastructure) as it's an entirely manual process to manage. For the most part, these items are segregated into their own namespaces so management with Flux seems easily achieved. 

However there are *some* resources (ex. `ConfigMap`, `Issuer`, etc.) that are shared across applications/services, relate to the infrastructure and are managed manually (not part of any application's chart). I classify these into the infrastructure group as well. In this case, there is *overlap* with the namespaces used by the application.

- Is such a configuration supported? 
- For overlapping namespaces, will Flux interfere with resources controlled by other means (ie. our ADO Helm deployments)? 
- Similarly, how does it interact with resources created by other tools? 
  - ex. I use `akv2k8s` which syncs `Secret` objects from Azure Keyvault into the application namespaces. Do I need to start adding additional labels to these? 
- How does Flux treat pre-existing resources when onboarding into an existing cluster?
  - Must I label everything ahead of time to ensure garbage collection skips it?
  - Or do I need to ensure I've got all existing resources configured within Flux before enabling it?
  - Or neither/something else?
 

Thanks! And apologies for the length, I thought of more questions as I was writing

**Comment by kingdonb** [comment, marked as answer]
Flux will not garbage collect any resources that it did not create. In earlier versions of flux, a label scheme was used for garbage collection, but we have moved on and the Flux Kustomization now uses an internal inventory to keep track of which resources it is synchronizing. Flux will not make any impact on resources which do not have a representation in any Source (GitRepository, etc.) that you have attached to the cluster.

You can mix resources in the same namespace and keep them "free-range" as together with others that are in Flux's management.

If you have resources in the cluster that you need Flux to patch, this is slightly tricky since Server-Side Apply made some changes in the way that resources are managed, you will need to ensure the resources to merge get this annotations treatment:

https://fluxcd.io/docs/faq/#how-to-patch-coredns-and-other-pre-installed-addons

Adding the (a)`ssa: merge` and (b)`prune: disabled` annotations as shown in the FAQ ensure that Flux will (a) not assume we have the resource's full definition in Git, that there may be some prior drift existing in the cluster before Flux's management was added, and we want to keep it; and (b) if Flux's config is ever removed from Git, then it should not be removed from the cluster, otherwise Flux will garbage collect resources when the `spec.prune` is true on the Kustomization, (but it can be blocked by adding this annotation.)

It would make sense to try managing your infrastructure with Flux as well, as having manually curated configmaps and hand-tooled Helm releases is dodgy work and automating it is Flux's main target market. But this can be done gradually, you can mix small Flux deployments in with pre-existing infrastructure and it does not need to be re-architected from scratch, not necessarily anyway.

In general, the expectation you should have is that Flux will only manage resources that you put under its management in Git. The default installation instructions for "bootstrapping" give you a Flux installation which manages only Flux itself, and has made room for configuration to be provided alongside which can be used to bootstrap additional configurations inside of Flux's management. Flux does not go outside of this bubble unless you ask it to.

**Reply by pinkfloydx33** [reply]
Ok cool, thanks, that makes sense. My biggest concern was about the garbage collection of non-flux managed resources. Maybe I missed it in the docs but it wasn't clear to me what would happen. My main goal with Flux would be management of my "infrastructure" helm charts and a set of manually created resources which I'm sick of managing myself. 

When onboarding Flux, what's the recommended way to deal with pre-existing resources, such as a helm chart, that I want Flux to start managing? Does it make sense to remove it first (if possible) and let Flux create/manage it anew? Or should it just sync right up, assuming no blunders in the configuration? 

I still worry about introducing a tool like this, despite any guarantees... I've considered spinning up a new cluster mirroring my current one and adding Flux to that as sort of a dry run just to be sure. I figure this would be a good way to ensure I get all of the helm configurations correct as well. Obviously you can't speak to my exact scenario, but do you think that sounds like overkill?

**Reply by stefanprodan** [reply]
Flux can take over all of your workloads and configs. For Helm you need to make sure that the Flux HelmRelease name and namespace matches the one install manually. As Kingdon said above, Flux does not touches resources it doesn’t manage, a resource applied with kubectl or Helm CLI will not be deleted if it’s not declared in Git.

**Reply by kingdonb** [reply]
> spinning up a new cluster mirroring my current one and adding Flux to that as sort of a dry run just to be sure

This is a great idea, and it's what I would recommend. You want to be sure you know all of the quirks before you introduce this type of automation to your existing production cluster. Things like setting `releaseName` on your `HelmReleases` that Stefan mentioned are examples of what you may trip over; in general the process of having Helm Controller take over an existing HelmRelease is smooth, but you want to try it in a non-production scenario first, especially if you are not already familiar with Helm Controller, and do become familiar with all the options presented in the [HelmRelease guide](https://fluxcd.io/docs/guides/helmreleases/) if you have not already.

If you have been manually curating helm releases you want to make sure you are familiar with `helm get` subcommands, especially `helm get values` since you will need to describe all of the values declaratively, this will help you to determine the current set of params, and as you can no longer depend on `--reuse-values` to aggregate values of past upgrades, this is another common need.

Taking the output from `helm get values` to compare with and build the `spec.values` that you'll set in your chart is the way to mitigate it. In Helm technically all values are secret, even though most values are likely simple config, at least some values are likely secrets and Helm Controller has a special way to handle secrets in values, for them you'll likely use `valuesFrom` with a secret applied perhaps with this approach: https://fluxcd.io/docs/guides/helmreleases/#refer-to-values-in-configmaps-generated-with-kustomize

There are definitely some tripping hazards and patterns to know here, and as with anything, you should use a staging environment to prove it works before trying any untested procedure or runbook on infrastructure or in production, there are also of course no firm guarantees based on how Flux is licensed, and nobody likes to have a surprise in production. 😓 😅 

If you are thinking about a dry-run first I definitely think that's a good idea, things always seem to turn out better for me when I don't expect them to go smoothly 😆
